File: v1/src/main/java/com/api/v1/doctors/controllers/DoctorController.java
package com.api.v1.doctors.controllers;

import com.api.v1.common.EmptyResponse;
import com.api.v1.doctors.requests.DoctorRegistrationDto;
import com.api.v1.doctors.responses.DoctorResponseDto;
import com.api.v1.doctors.services.DoctorManagementService;
import com.api.v1.doctors.services.DoctorRegistrationService;
import com.api.v1.doctors.services.DoctorRetrievalService;
import com.api.v1.doctors.services.DoctorUpdatingService;
import com.api.v1.people.requests.PersonUpdatingDto;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("api/v1/doctors")
public class DoctorController {


    private final DoctorManagementService managementService;
    private final DoctorRegistrationService registrationService;
    private final DoctorRetrievalService retrievalService;
    private final DoctorUpdatingService updatingService;

    public DoctorController(DoctorManagementService managementService,
                            DoctorRegistrationService registrationService,
                            DoctorRetrievalService retrievalService,
                            DoctorUpdatingService updatingService
    ) {
        this.managementService = managementService;
        this.registrationService = registrationService;
        this.retrievalService = retrievalService;
        this.updatingService = updatingService;
    }

    @PostMapping
    @Operation(summary = "Register a new doctor")
    public Mono<ResponseEntity<DoctorResponseDto>> register(@Valid @RequestBody DoctorRegistrationDto registrationDto) {
        return registrationService.register(registrationDto);
    }

    @PatchMapping("{doctorId}")
    @Operation(summary = "Update a new doctor")
    public Mono<ResponseEntity<EmptyResponse>> update(@PathVariable String doctorId, @Valid @RequestBody PersonUpdatingDto updatingDto) {
        return updatingService.update(doctorId, updatingDto);
    }

    @GetMapping("{id}")
    @Operation(summary = "Find a doctor by id")
    public Mono<ResponseEntity<DoctorResponseDto>> findById(@PathVariable String id) {
        return retrievalService.findById(id);
    }

    @GetMapping
    @Operation(summary = "Find all doctors")
    public ResponseEntity<Flux<DoctorResponseDto>> findAll() {
        return retrievalService.findAll();
    }

    @PatchMapping("{doctorId}/termination")
    @Operation(summary = "Terminate a doctor")
    public Mono<ResponseEntity<EmptyResponse>> terminate(@PathVariable String doctorId) {
        return managementService.terminate(doctorId);
    }

    @PatchMapping("{doctorId}/rehiring")
    @Operation(summary = "Rehire a doctor")
    public Mono<ResponseEntity<EmptyResponse>> rehire(@PathVariable String doctorId) {
        return managementService.rehire(doctorId);
    }
}

File: v2/src/main/kotlin/com/api/v2/doctors/controllers/DoctorController.kt
package com.api.v2.doctors.controllers

import com.api.v2.doctors.dtos.DoctorRegistrationDto
import com.api.v2.doctors.dtos.DoctorResponseDto
import com.api.v2.doctors.services.DoctorModificationService
import com.api.v2.doctors.services.DoctorRegistrationService
import com.api.v2.doctors.services.DoctorRetrievalService
import com.api.v2.doctors.services.DoctorTerminationService
import com.api.v2.people.dtos.PersonModificationDto
import jakarta.validation.Valid
import kotlinx.coroutines.flow.Flow
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("api/v2/doctors")
class DoctorController(
    private val registrationService: DoctorRegistrationService,
    private val retrievalService: DoctorRetrievalService,
    private val modificationService: DoctorModificationService,
    private val terminationService: DoctorTerminationService
) {

    @PostMapping
    @ResponseStatus(value = HttpStatus.CREATED)
    suspend fun register(@RequestBody registrationDto: @Valid DoctorRegistrationDto): DoctorResponseDto {
        return registrationService.register(registrationDto)
    }

    @GetMapping("{licenseNumber}")
    @ResponseStatus(value = HttpStatus.OK)
    suspend fun findByMedicalLicenseNumber(@PathVariable licenseNumber: String): DoctorResponseDto {
        return retrievalService.findByMedicalLicenseNumber(licenseNumber)
    }

    @GetMapping
    @ResponseStatus(value = HttpStatus.OK)
    suspend fun findAll(): Flow<DoctorResponseDto> {
        return retrievalService.findAll()
    }

    @PatchMapping("{licenseNumber}")
    @ResponseStatus(value = HttpStatus.NO_CONTENT)
    suspend fun modify(
        @PathVariable licenseNumber: String,
        @RequestBody modificationDto: @Valid PersonModificationDto
    ) {
        return modificationService.modify(licenseNumber, modificationDto)
    }

    @PatchMapping("{licenseNumber}")
    @ResponseStatus(value = HttpStatus.NO_CONTENT)
    suspend fun terminate(@PathVariable licenseNumber: String) {
        return terminationService.terminate(licenseNumber)
    }

}